src/index.js

```diff
@@ -10,21 +12,40 @@ export default function supernova(env) {
       properties,
       data,
     },
-    component: {
-      mounted(element) {
-        this.element = element;
-      },
-      render({ layout }) {
-        renderTree({
-          element: this.element,
-          layout,
-          model: this.model,
-          Theme,
-        });
-      },
-      resize() {},
-      willUnmount() {},
-      destroy() {},
+    component: () => {
+      //  const [focusNodes, setFocusNodes] = useState([]);
+      const [dataTree, setDataTree] = useState({});
+      //   const [storage, setStorage] = useState({});
+      const layout = useStaleLayout();
+      const model = useModel();
+      const element = useElement();
+      const rect = useRect();
+
+      const callRender = () => {
+        if (element && dataTree && layout) {
+          return renderTree({
```

Comment:
A function returned from `useEffect` is a 'cleanup' function that will be executed when the specified deps change, and when the component is unmounted/destroyed - you probably need to split this method for usePromise and useEffect

Reply:
think I want to use usePromise for all calls to render(), as they are likely to be async. Not needed now thought, and got into some infinite loops with it calling the rect one all the time.